### ISSUE TITLE
Issue 2: Implement a leakage and shortcut audit for all evaluation splits

### ISSUE THREAD

**Issue by Mati86** [AI-Security-Internships-2026/12-graph-based-cyber-physical-risk#23]
# Issue 2: Implement a leakage and shortcut audit for all evaluation splits

## Milestone: M1 — Reproducible Evaluation Foundation

---

## Problem Statement

Implement a reusable **split audit** that detects whether a train/test protocol accidentally creates an easy shortcut or an unrealistic evaluation condition.

The audit should be applied before model training and should describe exactly what the model is being asked to generalize to.

---

## Research Question Supported

**RQ1:** How strongly do common train/test partitioning strategies affect reported GNN-based ICS intrusion-detection performance?

---

### 1. Split Audit Module

Create a function/module such as:

```python
audit_split(train_df, test_df, graph_train=None, graph_test=None, ...)
```

or an equivalent dataset-aware implementation.

The audit must calculate, where applicable:

#### Identity and topology overlap

- source IP overlap,
- destination IP overlap,
- any-host/IP overlap,
- source-destination pair overlap,
- node overlap,
- edge/pair overlap,
- percentage of test nodes unseen during training,
- percentage of test edges/pairs unseen during training.

#### Attack coverage

For every attack subtype:

- training count,
- validation count,
- test count,
- number of unique source IPs,
- number of unique destination IPs,
- whether the attack subtype is evaluable under the chosen split,
- percentage of the attack taxonomy represented in the test set.

#### Temporal leakage

For chronological protocols:

- earliest/latest timestamp in each split,
- raw-record overlap,
- overlapping sliding windows,
- future graph edges visible to earlier predictions,
- duration represented by train/test periods.

#### Distribution shift

Report:

- normal/attack base rate,
- per-attack base rate,
- relative base-rate change,
- numeric feature drift summary,
- optionally PSI/KS statistics for key features,
- categorical distribution change where practical.

#### Shortcut warning rules

Generate warnings such as:

```text
WARNING: 10/13 attack classes absent from test set.
WARNING: 98% of test source IPs were seen in training.
WARNING: train/test sliding windows share raw observations.
WARNING: test graph contains edges occurring after the evaluated sample.
WARNING: one attack subtype is associated with one fixed source IP.
```

Warnings should be descriptive, not automatically interpreted as proof of leakage.

### 2. Split Audit Experiments

Run the split audit on at least:

1. SCADANet Track A.
2. SCADANet Track B.
3. SCADANet chronological split.
4. BATADAL static cross-validation setup.
5. BATADAL current chronological split.
6. The strict temporal protocols created later in Issue #5.

Issue #2 can initially produce results for 1–5 and later be updated when Issue #5 is completed.

### 3. Result Artifacts

Create:

```text
experiments/results/q1/split_audits/
```

Suggested files:

```text
scadanet_track_a_audit.json
scadanet_track_b_audit.json
scadanet_temporal_audit.json
batadal_static_audit.json
batadal_temporal_audit.json
```

Also create a consolidated CSV:

```text
split_audit_summary.csv
```

#### Table A1 — Evaluation protocol audit

| Dataset | Split | Host overlap | Pair overlap | Test attack types covered | New test nodes | New test edges | Temporal overlap |
|---|---|---:|---:|---:|---:|---:|---|
| SCADANet | Track A | ... | ... | 2/13 | ... | ... | N/A |
| SCADANet | Track B | ... | ... | 12/13 | ... | ... | N/A |
| SCADANet | Temporal | ... | ... | ... | ... | ... | ... |
| BATADAL | Temporal | N/A | N/A | binary | N/A | N/A | ... |

#### Table A2 — SCADANet per-attack split coverage

| Attack type | Unique source IPs | Train samples | Test samples Track A | Test samples Track B | Test samples Temporal |
|---|---:|---:|---:|---:|---:|
| vuln_scan | ... | ... | ... | ... | ... |
| tcp_syn_flood | ... | ... | ... | ... | ... |
| ... | ... | ... | ... | ... | ... |

This table is important because it makes the Track-A shortcut visible without relying on aggregate metrics.

#### Figure A1 — Attack-type coverage by split

A heatmap or presence/absence matrix:

```text
                  Track A    Track B    Temporal
vuln_scan            ✗          ✓          ✓
tcp_syn_flood        ✓          ...        ...
...
```

Preferred visual:

- rows = attack types,
- columns = evaluation protocols,
- cell value = number of test samples or binary coverage.

#### Figure A2 — Train/test overlap and novelty

A compact figure showing for each protocol:

- % seen hosts,
- % unseen hosts,
- % seen pairs,
- % unseen pairs.

#### Figure A3 — Base-rate shift across time

For SCADANet temporal split:

- x-axis = attack type,
- train prevalence,
- test prevalence,
- highlight `vuln_scan`.

This should make the already observed large `vuln_scan` shift immediately visible.

---

## Expected Research Result

The issue should not aim to make one model look good.

The expected result is a clear empirical answer to:

> What does each split actually test?

A successful result may show that:

- Track A measures only a small subset of the SCADANet attack taxonomy;
- Track B evaluates known behavior but has substantial host overlap;
- temporal splits introduce different base rates and topology novelty;
- overlapping BATADAL windows violate strict train/test independence.

---

## Acceptance Criteria

- [ ] A reusable split-audit function/module exists.
- [ ] Audit results are generated automatically from actual train/test data.
- [ ] All listed protocols are audited.
- [ ] Per-attack-type test coverage is reported.
- [ ] Host/node/pair overlap is reported where meaningful.
- [ ] Temporal/window overlap is reported where meaningful.
- [ ] Distribution/base-rate changes are quantified.
- [ ] Audit warnings are saved in machine-readable form.
- [ ] Table A1 and Table A2 data are exportable directly from committed results.
- [ ] Figures A1–A3 are generated from saved artifacts, not manually typed values.

---

## Do Not Do In This Issue

Do not:

- change model architecture,
- tune thresholds,
- delete inconvenient attack categories,
- interpret overlap alone as causation,
- use model test performance to define the split.

This issue is about understanding evaluation conditions before interpreting model performance.
